text/0000-mut-pattern-shorthand.md

```diff
@@ -0,0 +1,569 @@
+As such, the RFC's author is not aware of any prior art.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+None as of yet.
+
+# Future work
+[future work]: #future-work
+
+In this section, we consider some possible future work that could be done
+but shouldn't necessarily be done.
+
+## Structs and fields
+
+It would be possible to extend the shorthand syntaxes proposed in this RFC with:
+
+```rust
+pat
+: // other productions
+| MUT path_expr '{' pat_struct '}'
+| MUT path_expr '(' pat_tup ')'
+```
+
+This would allow you to transform the following:
+
+```rust
+struct Person<'a> { age: usize, hobby: &'a str, pocket: [&'a str; 3] }
+
+// NOTE: Full mutability:
+let Person { mut age, mut hobby, pocket: [mut x, mut y, mut z] }
+  = Person { age: 1, hobby: "designing Rust"
+           , pocket: ["key", "phone", "headphones"] };
+
+struct Point(usize, usize);
+
+let Foo(mut x, mut y) = Foo(1, 2);
+```
+
+into:
+
+```rust
+struct Person<'a> { age: usize, hobby: &'a str, pocket: [&'a str; 3] }
+
+// NOTE: Full mutability:
+let mut Person { age, hobby, pocket: [x, y, z] }
+  = Person { age: 1, hobby: "designing Rust"
+           , pocket: ["key", "phone", "headphones"] };
+
+let mut Foo(x, y) = Foo(1, 2);
+```
+
+## Or-patterns
+
+Another extension could be to allow `mut PAT | .. | PAT`. This could be done by
+redefining [`pats_or`](https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y#L983-L986)
+as:
+
+```rust
+pats_or
+: MUT pats_or_old
+| pats_or_old
+;
+
+pats_or_old
+: pat
+| pats_or_old '|' pat
+;
+```
+
+This also applies to `let` bindings, `if let`, and `while let` and not
+just `match` arms.
+
+Note that `mut PAT | .. | PAT` is interpreted as `mut (PAT | .. | PAT)` and
+**not** `mut PAT | (.. | PAT)`.
+
+From a user's perspective, this would allow you to replace:
+
+```rust
+enum E<T> { A(T), B(T), C(T) }
+
+match E::A(1) {
+    E::A(mut x) | E::B(mut x) => { x = 3; },
+    E::C(mut x) => { x = 2; }
+}
+```
+
+with:
+
+```rust
+enum E<T> { A(T), B(T), C(T) }
+
+match E::A(1) {
+    mut E::A(x) | E::B(x) => { x = 3; },
+    mut E::C(x) => { x = 2; }
+}
+```
+
+However, this would also interfere with `mut x @ Foo(_) | mut x @ Bar(_)`.
+
+### Rationale for precedence in `mut PAT | .. | PAT`
+
+With this extension, you would interpret `mut PAT | .. | PAT` as
+`mut (PAT | .. | PAT)` instead of `mut PAT | (.. | PAT)`.
+
+We argue that this is correct because if it would be interpreted in the second
+way, then you'd get the following:
+
+```rust
+enum E<T> { A(T), B(T), C(T) }
+
+match E::A(1) {
+    E::A(mut x) | E::B(x) => { x = 3; },
+    E::C(mut x) => { x = 2; }
+}
+```
+
+which results in an error:
+
+```rust
+error[E0409]: variable `x` is bound in inconsistent ways within the same match arm
+ --> src/main.rs:6:24
+  |
+6 |     E::A(mut x) | E::B(x) => { x = 3; },
+  |              -         ^ bound in different ways
+  |              |
+  |              first binding
+```
+
+## Reversed polarity with `immut`
```

Comment:
We could use `const` (as in `*mut T` vs `*const T`) instead of introducing yet another keyword. (I'm :-1: to allowing reversed polarity anyway).

Reply:
So the point of this section is to say that we should not introduce `immut` ;) So I agree entirely with you here that we shouldn't do it.